v2/internal/util/match/string_matcher.go

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package match
+
+import (
+	"fmt"
+)
+
+// StringMatcher is an interface implemented by predicates used to test string values
+type StringMatcher interface {
+	fmt.Stringer
+
+	// Matches returns true if the provided value is matched by the matcher
+	Matches(value string) bool
+	// WasMatched returns nil if the matcher had a match, otherwise returning a diagnostic error
+	WasMatched() error
+	// IsRestrictive returns true if the matcher is populated and will restrict matches
+	IsRestrictive() bool
+}
+
+// NewStringMatcher returns a matcher for the specified string
+// Different strings may return different implementations:
+// o If the string contains '*' or '?' a globbing wildcard matcher
+// o Otherwise a case-insensitive literal string matcher
+func NewStringMatcher(matcher string) (StringMatcher, error) {
+	if HasWildCards(matcher) {
+		return NewGlobMatcher(matcher)
+	}
+
+	return NewLiteralMatcher(matcher), nil
+}
+
+// NewStringMultiMatcher returns a matcher for the specified string
+// Different strings may return different implementations:
+// o If the string contains ';', a multi-matcher of sub-matches, one for each item in the string separated by ';'
+// o If the string contains '*' or '?' a globbing wildcard matcher
+// o Otherwise a case-insensitive literal string matcher
+func NewStringMultiMatcher(matcher string) (StringMatcher, error) {
+	if HasMultipleMatchers(matcher) {
+		return NewMultiMatcher(matcher)
+	}
+
+	return NewStringMatcher(matcher)
+}
```

Comment:
Do we really need multiple entry points like this? Under what circumstances would a multimatcher be a problem?

Reply:
The `--crd-patterns` cmdline argument was originally documented as comma delimited, and this felt weird: `"foo/*;bar/*,baz/*"` -- but I suppose we could switch it to be `;` delimited in which case we could have a single path here? Thoughts?

Reply:
I'd favour changing `MultiMatcher` so that it accepted either `;` or `,` (but not both), giving users flexibility without allowing things that would be confusing (e.g. should `"foo/*;bar/*,baz/*"` be treated as `"foo/*` + `bar/*,baz/*"` or `"foo/*;bar/*` + `baz/*"` or `"foo/*` + `bar/*` + `baz/*"`)

I think then we could go back to a single entry point. 

I don't think it matters if we document the patterns as `,` separated and then happen to accept `;` as well - it's a case of Postel's law.

Reply:
Let me look into this some more and see what makes sense here. Will update shortly

Reply:
Ok so I dug into this a bit more and here's what I've got:

Here's the codepath as I have it locally:
1. User specifies `--crd-pattern=a,b,c,d`. This is parsed by [string_slice_flags](https://github.com/Azure/azure-service-operator/blob/main/v2/internal/util/flags/string_slice_flags.go) into a `[]string`.
2. This flows into a function `filterCRDsByPatterns` `in `crdmanagement` (new in my local) which translates each user specified pattern into a `match.StringMatcher`. So it's a `[]string` -> `[]match.StringMatcher` conversion.
3. Each matcher is checked individually, and each CRD it matches is cataloged.
4. An error is raised if any matcher doesn't match anything

We could re-use the `MultiMatcher` implementation and go from `[]string` -> `match.MultiMatcher` (or possibly from `string` -> `match.MultiMatcher`, but this presents one challenge:

1. I want to log which pattern matched each CRD.
    a. This could probably be fixed by extending the `match.StringMatcher` interface to have some sort of `bool, string` result where it included textual details about which pattern matched each string, although that interface extension is a bit awkward for most of the other `match` implementations because they only ever can be matched by a single pattern so it's sorta obvious.


I'll investigate changing this and see how it goes.

Reply:
Notes: What I have is pretty good but 1 more change which is to return a result type rather than a bool directly, and avoid 2x methods.

Reply:
Fixed